Added binary input option using Kryo

Sahasrara opened the pull request.

Commit:
Added binary input option using Kryo

jexp commented on src/main/java/org/neo4j/shell/tools/imp/format/kryo/KryoSerializationTypes.java:
```diff
@@ -0,0 +1,12 @@
+/**
+ * Created by efulton on 5/5/16.
+ */
+public enum KryoSerializationTypes {
```
could we make these enums strings? so that someone else with another tool could import the data too even if they don't have shell tools

jexp replied:
@Sahasrara what do you think?

Sahasrara replied:
Sure, I'll just use the .name() for the enums

jexp commented on src/main/java/org/neo4j/shell/tools/imp/format/kryo/KryoWriter.java:
```diff
@@ -0,0 +1,113 @@
+        // WRITE LABEL
+        kryo.writeObject(output, relationship.getType().name());
+        // WRITE PROPERTIES
+        Map<String, Object> properties = relationship.getAllProperties();
```
should it only do that for non-empty maps? perhaps it's faster to check if the rel has a property or not? 
perhaps also generalize it over property-containers?

Sahasrara replied:
Always having a map (empty or not) saves us from writing out an extra flag indicating whether or not there is a map.  I think that the way kryo works, the difference in size is pretty much negligible in practice, but I can check that to make sure.

jexp replied:
You have to put the counter between the two streams.

jexp commented on src/main/java/org/neo4j/shell/tools/imp/format/kryo/KryoReader.java:
```diff
@@ -0,0 +1,142 @@
+                    // Add Relationship to Database
+                    RelationshipType relationshipType = relationshipLabel != null ?
+                        DynamicRelationshipType.withName(relationshipLabel) : this.relType;
+                    Node from = gdb.getNodeById(cache.get(startNodeId));
```
not sure we really have to cache this

jexp commented on src/main/java/org/neo4j/shell/tools/imp/format/kryo/KryoReader.java:
```diff
@@ -0,0 +1,142 @@
+                    Node from = gdb.getNodeById(cache.get(startNodeId));
+                    Node to = gdb.getNodeById(cache.get(targetNodeId));
+                    Relationship relationship = from.createRelationshipTo(to, relationshipType);
+                    relationship.setProperty("id", id);
```
this changes the original data, I think this should be an option to keep old rel-ids and node-id's around

Sahasrara replied:
Uh oh, which line does that?  I thought I had preserved everything that could be preserved.

jexp replied:
What I meant: if you export and import a graph, you don't get the same data, as the import adds these id's. (which might even override existing business-id fields).

That's why I would want this to be an option for import-binary.

Sahasrara replied:
So in the case that a user requests we preserve the id's, I should keep things as is.  IF the user does not want to preserve the id's, I should NOT add the id's to the nodes/rels.  

Am I understanding correctly?  If so, I'll go ahead and add a flag for that as you suggest.  I just wanted to make sure I understand correctly before I do.

jexp replied:
the default should be to not add the additional id field

Sahasrara replied:
OHHH shoot.  I thought i was setting the node id.  My mistake.  I'll fix that.

jexp commented on src/main/java/org/neo4j/shell/tools/imp/format/kryo/KryoReader.java:
```diff
@@ -0,0 +1,142 @@
+
+                    // Add Node to Database
+                    Node node = gdb.createNode(labels.toArray(new Label[labels.size()]));
+                    node.setProperty("id", id);
```
see rel-id comment

Sahasrara replied:
I thought that:

`node.setProperty("id", id);`

would take care of that?

jexp commented on src/main/java/org/neo4j/shell/tools/imp/ImportKryoApp.java:
```diff
@@ -0,0 +1,63 @@
+    {
+        addOptionDefinition("i", new OptionDefinition(OptionValueType.MUST, "Input Binary file"));
+        addOptionDefinition("b", new OptionDefinition(OptionValueType.MUST, "Batch Size default " + DEFAULT_BATCH_SIZE));
+        addOptionDefinition("r", new OptionDefinition(OptionValueType.MUST, "Default Relationship-Type otherwise edge attribute 'label' " + DEFAULT_REL_TYPE));
```
then it should not be a MUST

Sahasrara replied:
I'll make -r a "MAY", but I had just copied it from the xml app.

jexp commented on src/main/java/org/neo4j/shell/tools/imp/ImportKryoApp.java:
```diff
@@ -0,0 +1,63 @@
+        addOptionDefinition("i", new OptionDefinition(OptionValueType.MUST, "Input Binary file"));
+        addOptionDefinition("b", new OptionDefinition(OptionValueType.MUST, "Batch Size default " + DEFAULT_BATCH_SIZE));
+        addOptionDefinition("r", new OptionDefinition(OptionValueType.MUST, "Default Relationship-Type otherwise edge attribute 'label' " + DEFAULT_REL_TYPE));
+        addOptionDefinition("c", new OptionDefinition(OptionValueType.NONE, "Use a node-cache that overflows to disk, necessary for very large imports"));
```
not sure we need the cache at all

jexp commented on src/main/java/org/neo4j/shell/tools/imp/format/kryo/KryoReader.java:
```diff
@@ -0,0 +1,142 @@
+                    tx.increment();
+                    reporter.update(0, 1, properties.size());
+                } else {
+                    throw new RuntimeException("Encountered corrupt serialization");
```
could we continue even if the serialization is partially corrupt?

Sahasrara replied:
The only scenario where we could continue, is if there was a superfluous unknown TYPE in the file followed immediately by another TYPE.  I can make this a log statement, but I'm not sure this is a recoverable scenario in most circumstances.

jexp commented on src/main/java/org/neo4j/shell/tools/imp/ExportKryoApp.java:
```diff
@@ -0,0 +1,72 @@
+
+        String fileName = parser.option("o", null);
+
+        com.esotericsoftware.kryo.io.Output output = new com.esotericsoftware.kryo.io.Output(new FileOutputStream(fileName));
```
does it make sense to zip a kryo file? esp. with the many repeated strings for property-keys, labels and rel-types and also the same node-id's repeated?

I wonder if we should write a translation table at the beginning for property-keys, labels and rel-types -> int and use that throughout

Sahasrara replied:
Zipping makes a lot of sense.  The only reason I didn't use Inflator/Deflator input/output is it obscures the progress reporting since progress reports are based on size.  I think zipping it after the fact might make most sense since it'll take care of efficient compression for you.  I'm totally open to suggestion there though.  When I tried inflator/deflator on my database 300MB became 100MB.

jexp replied:
I would just wrap your stream into a GZipOutputStream. We could make this an option, or base it on the extension (e.g. if a user provides foo.gz as a name we zip it) but we can also zip it in general (that would be my preference).

Sahasrara replied:
So I did something like that to begin with (except with inflator/deflator input/output streams).  This obscured the actual file size during the reporting progress so we'd get well over 100% printed out in the progress reporter.

Would you prefer this WITH progress reporting:

Write:
output binary file
compress binary file

Read:
decompress binary file
input binary file

OR this WITHOUT progress reporting:

Write:
output compressed binary file

Read
input compressed binary file

Let me know what you think, or if that doesn't make sense I can try to explain it better.

Sahasrara replied:
nevermind on this.  I was mistaken about not being able to calculate percent complete.  I just confused myself ¯_(ツ)_/¯

jexp commented:
A general note. What do you think about making the format an option, as you already encapsulated it in the writer / reader? So you could do `-format kryo´ or`-format packstream`or`-format cbor` ? And defaulting to one?

Sahasrara commented:
That sounds like a good idea to me.  Adding the format flag to open the door for future binary formats makes sense to me.  I'll add that in with the other stuff I'll fix for this pull request.  I'll send in another request so that I do everything at once.  I just want to make sure I understand some of the other stuff before I start the next revision.

Commit:
Addressing comments from pull request

 - Kryo enums are stored as strings
 - Node ID is not saved as an additional attribute for imported nodes/rels.  That was a mistake.
 - Compression
 - CLI options are properly labeled as required or not
 - Added groundwork for different binary formats

Commit:
Removing commented code

jexp commented:
@Sahasrara is it ready to merge or are you still working on it?

Sahasrara commented:
That's all I had for now so if you're OK with it, it's good to go.

jexp commented:
Merged manually into 2.3 and 3.0 thanks so much !
